### PR TITLE
[Merged by Bors] - feat: `ring_exp` tactic, subtraction in `ring`

### PR DESCRIPTION
**Pull request by digama0**
Another complete rewrite. This implements one `ring` to rule them all: it incorporates both `ring_exp` and `ring` behaviors, now under the name `ring`. (`ring_exp` had some small (1.4x) performance issues that prevented it from being used by default; I'm hoping that those issues are fixed now, and we can revisit later if it becomes an issue.)

### CODE REVIEW

File: Mathlib/Tactic/Ring.lean [hunk 1]
```diff
@@ -1,426 +1,873 @@
...
+/--
+A monomial, which is a product of powers of `ExBase` expressions,
+terminated by a (nonzero) constant coefficient.
+-/
+inductive ExProd : ∀ {α : Q(Type u)}, Q(CommSemiring $α) → (e : Q($α)) → Type
+  /-- A coefficient `value`, which must not be `0`. `e` is a raw int cast. -/
+  | const (value : ℤ) : ExProd sα e
+  /-- A product `x ^ e * b` is a monomial if `b` is a monomial. Here `x` is a `ExBase`
+  and `e` is a `ExProd` representing a monomial expression in `ℕ` (it is a monomial instead of
+  a polynomial because we eagerly normalize `x ^ (a + b) = x ^ a + x ^ b`.) -/
```
Comment (semorrison) [inline Mathlib/Tactic/Ring.lean:147]:
```suggestion
  a polynomial because we eagerly normalize `x ^ (a + b) = x ^ a * x ^ b`.) -/
```

**Comment by semorrison** [comment]
What about goals like
```
example {α} [CommSemiring α] (x y : α) : (x + y)^2 = x^2 + 2 • x * y + y^2 := by ring
```
which should also count as "in the language of commutative rings"?

Presumably it is easier to first `simp` away all the `•`s, and then proceed as is, rather than adding an `evalSmul` branch, etc.

**Comment by semorrison** [comment]
bors d+

**Comment by digama0** [comment]
bors merge

**Pull request closed without merging**